### PR TITLE
Initial support for OpenRISC 1000

### PR DESCRIPTION
**Pull request by samueldr**
## Motivation

https://github.com/NixOS/mobile-nixos/issues/183#issuecomment-659865562

This will be required to build `crust` a firmware for the Pinephone.

This is heavily inspired from @cleverca22's vc4 PR.

## Status

This currently builds GNU Hello

```
 ~/tmp/nixpkgs/nixos-unstable $ nix-build -A pkgsCross.or1k.hello
/nix/store/fisgprfv00mwfacxilj5r07v0w5ir2s0-hello-2.10-or1k-elf

 ~/tmp/nixpkgs/nixos-unstable $ file result/bin/hello 
result/bin/hello: ELF 32-bit MSB executable, OpenRISC, version 1 (SYSV), statically linked, not stripped
```

* * *

cc @noneucat

### CODE REVIEW

**Comment by Ericson2314** [comment]
If the stock newlib works, and doesn't require `CC=cc`, that sounds worth perusing to me. IMO `CC=cc` to me is a bad open as to that fork's maintainability: though that particular would be easy to fix, it makes me wonder what else is going on.

**Comment by samueldr** [comment]
Both fail with the same issue, so I don't know if any of them work yet. Noting here [that `CC=cc` is verbatim from the mainline newlib derivation](https://github.com/NixOS/nixpkgs/blob/36adeebde6cc1035f2f3a58a131376b2294f2f12/pkgs/development/misc/newlib/default.nix#L14-L17).

**Comment by noneucat** [comment]
I think we may need a newer version of `binutils` to build this toolchain:

https://github.com/bminor/binutils-gdb/blob/binutils-2_31-branch/cpu/or1k.opc -> no ha() implemented
https://github.com/bminor/binutils-gdb/blob/binutils-2_32-branch/cpu/or1k.opc -> ha() implemented at line 290

**Comment by samueldr** [comment]
Oh hi!

This is using parts of #89793, namely the changes needed to build binutils 2.34.

The changes *I think* end up not being a no-op. I'm not positive as to why yet.

Though before delving into that, if it even matters, I'd like people's opinion on that.

though:

```
 ~/tmp/nixpkgs/nixos-unstable $ nix-build -A pkgsCross.or1k.hello
...
 ~/tmp/nixpkgs/nixos-unstable $ file result/bin/hello 
result/bin/hello: ELF 32-bit MSB executable, OpenRISC, version 1 (SYSV), statically linked, not stripped
```

I guess that could mean, @noneucat, that we can move forward into making a crust derivation. As time permits :).

File: pkgs/development/tools/misc/binutils/default.nix [hunk 1]
```diff
@@ -33,59 +37,80 @@ let
...
     # un-break features so linking against musl doesn't produce crash-only binaries
-    ./0001-x86-Add-a-GNU_PROPERTY_X86_ISA_1_USED-note-if-needed.patch
-    ./0001-x86-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED.patch
-    ./0001-x86-Properly-add-X86_ISA_1_NEEDED-property.patch
-  ] ++ lib.optional stdenv.targetPlatform.isiOS ./support-ios.patch;
+    ./patches/2.31/0001-x86-Add-a-GNU_PROPERTY_X86_ISA_1_USED-note-if-needed.patch
+    ./patches/2.31/0001-x86-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED.patch
+    ./patches/2.31/0001-x86-Properly-add-X86_ISA_1_NEEDED-property.patch
+  ]
+  ++ lib.optional stdenv.targetPlatform.isiOS ./support-ios.patch
```
Comment (Ericson2314) [inline pkgs/development/tools/misc/binutils/default.nix:103]: was this one not moved?

**Review by Ericson2314 — APPROVED** [review]
Just one question on a darwin patch. Otherwise looks good!

File: pkgs/development/tools/misc/binutils/default.nix (hunk 1 above)
Comment (samueldr) [reply pkgs/development/tools/misc/binutils/default.nix:103]: It wasn't edited in #89793.

I left it as it was as I do not have the knowledge to test it.

We could move it to 2.31.1, and copy into 2.34 without changes, but I'm almost assuming the patch might not apply?

**Pull request merged**